File: RL_Maggie/Code/09_flex_label.R
## ------------------------------------------------------------------
## 04b_flex_label.R  – Derive “flex” vs “non-flex” label
## ------------------------------------------------------------------
## Prerequisites:
##   • full_trial_data_learners   (from 04_trial_merge_clean.R)
##   • correct column is 1/0      (already enforced earlier)
## ------------------------------------------------------------------

library(dplyr)
library(tidyr)
library(stringr)

# 1. accuracy profile: 6 blocks (1–6) × 2 pair types  --------------------
acc_profile <- full_trial_data_learners %>%
  filter(block %in% 1:6) %>%                     # discard extra learning trials (if any)
  group_by(participant, pair_type, block) %>%
  summarise(acc = mean(correct, na.rm = TRUE), .groups = "drop") %>%
  unite(cell, pair_type, block, sep = "_") %>%   # e.g. "reversed_4"
  pivot_wider(names_from = cell, values_from = acc)

# 2. person-mean centre ---------------------------------------------------
acc_mat          <- as.matrix(acc_profile[ , -1])              # strip participant id
row_means        <- rowMeans(acc_mat, na.rm = TRUE)
acc_mat_centred  <- sweep(acc_mat, 1, row_means)               # subtract

# 3. k-means (k = 2) ------------------------------------------------------
set.seed(2025)
km <- kmeans(acc_mat_centred, centers = 2, nstart = 50)

# 4. decide which cluster is “flex” --------------------------------------
#   Calculate mean (reversed – non-reversed) accuracy over blocks 4–6
gap_by_cluster <- sapply(1:2, function(cl){
  idx <- km$cluster == cl
  rev_cols  <- str_subset(colnames(acc_mat_centred), "^reversed_[4-6]$")
  non_cols  <- str_subset(colnames(acc_mat_centred), "^non-reversed_[4-6]$")
  mean(acc_mat_centred[idx , rev_cols] - acc_mat_centred[idx , non_cols], na.rm = TRUE)
})

flex_cluster <- which.max(gap_by_cluster)        # larger gap = more flexible

# 5. add label back to your master data frame ----------------------------
flex_labels <- acc_profile %>%
  mutate(flex_label = ifelse(km$cluster == flex_cluster, "flex", "non-flex")) %>%
  select(participant, flex_label)

# You can now merge with full_trial_data_learners, e.g.:
full_trial_data_learners <- full_trial_data_learners %>%
  left_join(flex_labels, by = "participant")

View(full_trial_data_learners)
# ------------------------------------------------------------------
# RESULT:  full_trial_data_learners now contains a column `flex_label`
#          (values: "flex" / "non-flex") exactly like the paper’s.
# ------------------------------------------------------------------

File: RL_Maggie/python/09_flex_label.py
import pandas as pd
from sklearn.cluster import KMeans
import importlib

merge_mod = importlib.import_module('04_trial_merge_clean')
full_trial_data_learners = merge_mod.full_trial_data_learners

acc_profile = (full_trial_data_learners[full_trial_data_learners['block'].between(1, 6)]
               .groupby(['participant', 'pair_type', 'block'])['correct']
               .mean().reset_index()
               .assign(cell=lambda d: d['pair_type'] + '_' + d['block'].astype(str))
               .pivot(index='participant', columns='cell', values='correct'))

acc_mat = acc_profile.fillna(acc_profile.mean()).to_numpy()
row_means = acc_mat.mean(axis=1, keepdims=True)
acc_mat_centred = acc_mat - row_means

km = KMeans(n_clusters=2, n_init=50, random_state=2025).fit(acc_mat_centred)

rev_cols = [c for c in acc_profile.columns if c.startswith('reversed_') and c.split('_')[1] in ['4', '5', '6']]
non_cols = [c for c in acc_profile.columns if c.startswith('non-reversed_') and c.split('_')[1] in ['4', '5', '6']]

gap_by_cluster = []
for cl in range(2):
    idx = km.labels_ == cl
    gap_by_cluster.append((acc_mat_centred[idx][:, [acc_profile.columns.get_loc(c) for c in rev_cols]].mean() -
                           acc_mat_centred[idx][:, [acc_profile.columns.get_loc(c) for c in non_cols]].mean()))
flex_cluster = int(pd.Series(gap_by_cluster).idxmax())

flex_labels = pd.DataFrame({
    'participant': acc_profile.index,
    'flex_label': ['flex' if lab == flex_cluster else 'non-flex' for lab in km.labels_]
})

full_trial_data_learners = full_trial_data_learners.merge(flex_labels, on='participant', how='left')
